torch/_vmap_internals.py

```diff
@@ -101,21 +101,21 @@ def _unwrap_batched(
     if isinstance(batched_outputs, Tensor):
         out_dim = out_dims_as_tuple[0]
         return torch._remove_batch_dim(batched_outputs, vmap_level, batch_size, out_dim)  # type: ignore[return-value]
-    return tuple(torch._remove_batch_dim(out, vmap_level, batch_size, out_dim)
+    return tuple((torch._remove_batch_dim(out, vmap_level, batch_size, out_dim) if out is not None else None)
```

Comment:
Hmm, I don't know if this is the semantics we want for vmap. The rough contract for vmap is, in the absence of mutation, running `vmap(f)(x)` returns a stacked version of `[f(i) for i in x.unbind()]`.

Could you explain a bit more about the context behind this change?

Reply:
The interesting thing is that JAX's vmap allows None to pass through, so that's an argument for why we should do the same

Reply:
I messed up the PR ordering actually - this is meant to enable the previous PR, which was to allow vmap to directly wrap the call to the engine, which may sometimes return None when certain inputs are unused.

There doesn't seem to be an easy workaround since we'd either have to have a wrapper that:
 - either materializes the None into a dummy (zero?) tensor, then detect those and turn those back into `None` before returning
 - or has some side effect to keep track of where to insert the `None`s back.

Previously the application was to compute Jacobian in a vectorized way, and that was fine because we wanted to materialize the None into zeros anyway, but here in the more generic case we don't.

> Hmm, I don't know if this is the semantics we want for vmap. The rough contract for vmap is, in the absence of mutation, running vmap(f)(x) returns a stacked version of [f(i) for i in x.unbind()].

On the surface this contract seems a little restrictive. I feel like it would be convenient to allow users to return non-tensors and have those pass through. Is the reasoning behind this restriction to prevent user error?

Reply:
Interestingly, in JAX, non-tensors don't actually pass through (only None does). For example, if the function is
```
def f(x):
  return 0
  
vmap(f)(np.ones([10]))
```
then the result is a Tensor/array of 10 zeros.

Reply:
I'm still concerned about allowing None to pass through. In the vmap API [one passes an `out_dims` argument](https://github.com/facebookresearch/functorch/blob/1813f4a6d86ff9783996ac25b21c413511046793/functorch/_src/vmap.py#L172) that says "where the batch dim should appear in the output". The `None` object has no concept of dimensions though. Could you open an issue over at the [functorch repo](https://github.com/facebookresearch/functorch) please? I'm curious what other people think.

Here's a workaround for now: instead of modifying the vmap API, we can have our own "autograd-safe" vmap that allows Nones to pass through. It would probably look something like:

```py
def _vmap(func: Callable, in_dims: in_dims_t = 0, out_dims: out_dims_t = 0) -> Callable:
    @functools.wraps(func)
    def wrapped(*args, **kwargs):
        _check_out_dims_is_int_or_int_pytree(out_dims, func)
        batch_size, flat_in_dims, flat_args, args_spec = _process_batched_inputs(in_dims, args, func)
        vmap_level = _vmap_increment_nesting(batch_size)
        try:
            batched_inputs = _create_batched_inputs(flat_in_dims, flat_args, vmap_level, args_spec)
            batched_outputs = func(*batched_inputs, **kwargs)
            return _unwrap_batched(batched_outputs, out_dims, vmap_level, batch_size, func, allow_none_passthrough=True)
        finally:
            _vmap_decrement_nesting()
    return wrapped
```